Source/utils/paths.cpp

```diff
@@ -1,6 +1,11 @@
 #include "utils/paths.h"
 
-#include <optional>
+#if !defined(__clang__) && defined(__GNUC__) && __GNUC__ < 7

```

Comment:
Not thrilled about this but if we're gonna do it we should use the `#if defined(__cpp_lib_optional)` feature detection macro.
https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros

Reply:
That doesn't seem to work, `__cpp_lib_optional` is not defined only -after- including `<optional>`, so that won't work if that header is not present.

Reply:
Ah, make sense. Let's use `__has_include` then; something like this:

```c++
#ifdef __has_include
#if defined(__cplusplus) && __cplusplus >= 201606L && __has_include(<optional>)
#define HAVE_STD_OPTIONAL
#elif __has_include(<experimental/optional>)
#define HAVE_EXPERIMENTAL_OPTIONAL
#endif
#endif

#if defined(HAVE_STD_OPTIONAL)
#include <optional>
#elif defined(HAVE_EXPERIMENTAL_OPTIONAL)
#include <experimental/optional>
#define optional experimental::optional
#else
#error "Missing support for <optional> or <experimental/optional>
#endif 
```

and please place into a separate header

Reply:
Ah nice, I didn't know about `__has_include`. Fancy.

Reply:
Now I feel silly for not remember reading about that yesterday...

Reply:
That code doesn't work well with MSVC though: https://ci.appveyor.com/project/AJenbo/devilutionx/builds/38864222 It does support `<optional>`.